File: trip-kit-booking-ui/src/main/java/com/skedgo/android/tripkit/booking/ui/viewmodel/ExternalViewModel.java
package com.skedgo.android.tripkit.booking.ui.viewmodel;

import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;
import android.os.Bundle;
import android.support.annotation.VisibleForTesting;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.skedgo.android.tripkit.booking.ExternalFormField;

import javax.inject.Inject;

import rx.Observable;
import rx.subjects.PublishSubject;
import skedgo.common.BaseViewModel;

public class ExternalViewModel extends BaseViewModel {

  private final ObservableBoolean showWebView;
  private final ObservableField<String> url;
  private final PublishSubject<String> publishSubjectNextUrl = PublishSubject.create();
  private ExternalFormField externalFormField;

  @Inject ExternalViewModel() {
    showWebView = new ObservableBoolean(false);
    url = new ObservableField<>();
  }

  public ObservableBoolean showWebView() {
    return showWebView;
  }

  public ObservableField<String> url() {
    return url;
  }

  public Observable<String> nextUrlObservable() {
    return publishSubjectNextUrl.asObservable();
  }

  public void handleArgs(Bundle args) {
    externalFormField = args.getParcelable("externalFormField");
    if (externalFormField != null) {
      url.set(externalFormField.getValue());
    }
  }

  public WebViewClient webViewClient() {
    return new WebViewClient() {
      @Override
      public boolean shouldOverrideUrlLoading(WebView webView, String url) {
        return handleCallback(url);
      }

      @Override
      public void onPageFinished(WebView view, final String url) {
        if (url.startsWith(externalFormField.getDisregardURL()) &&
            externalFormField.getValue().startsWith(externalFormField.getDisregardURL())) {
          publishSubjectNextUrl.onNext(externalFormField.getNextURL());
        } else {
          showWebView.set(true);
        }
      }
    };
  }

  @VisibleForTesting
  boolean handleCallback(String webUrl) {
    if (webUrl.startsWith(externalFormField.getDisregardURL())) {
      publishSubjectNextUrl.onNext(externalFormField.getNextURL());
      return false;
    } else {
      showWebView.set(true);
      this.url.set(webUrl);
    }
    return true;
  }

}

File: trip-kit-booking-ui/src/main/java/com/skedgo/android/tripkit/booking/ui/viewmodel/ExternalViewModel.kt
package com.skedgo.android.tripkit.booking.ui.viewmodel

import androidx.databinding.ObservableBoolean
import androidx.databinding.ObservableField
import android.os.Bundle
import androidx.annotation.VisibleForTesting
import android.webkit.WebView
import android.webkit.WebViewClient
import com.skedgo.android.tripkit.booking.ExternalFormField
import com.skedgo.android.tripkit.booking.ui.activity.KEY_EXTERNAL_FORM
import rx.Observable
import rx.subjects.PublishSubject
import javax.inject.Inject

class ExternalViewModel @Inject internal constructor() : DisposableViewModel() {

  private val publishSubjectNextUrl = PublishSubject.create<String>()
  @VisibleForTesting var externalFormField: ExternalFormField? = null

  val showWebView: ObservableBoolean = ObservableBoolean(false)
  val url: ObservableField<String> = ObservableField()
  val nextUrlObservable: Observable<String> = publishSubjectNextUrl.asObservable()


  fun handleArgs(args: Bundle) {
    externalFormField = args.getParcelable<ExternalFormField>(KEY_EXTERNAL_FORM)
    externalFormField?.let { url.set(it.value) }
  }

  fun webViewClient(): WebViewClient {
    return object : WebViewClient() {
      override fun shouldOverrideUrlLoading(webView: WebView, url: String): Boolean {
        return handleCallback(url)
      }

      override fun onPageFinished(view: WebView, url: String) {
        with(externalFormField!!) {
          if (url.startsWith(disregardURL) && value!!.startsWith(disregardURL)) {
            publishSubjectNextUrl.onNext(nextURL)
          } else {
            showWebView.set(true)
          }
        }
      }
    }
  }

  @VisibleForTesting
  fun handleCallback(webUrl: String): Boolean {
    return when {
      webUrl.startsWith(externalFormField!!.disregardURL) -> {
        publishSubjectNextUrl.onNext(externalFormField!!.nextURL)
        false
      }
      else -> {
        showWebView.set(true)
        url.set(webUrl)
        true
      }
    }
  }
}